Add support for loading synthetic block models, and using vanilla's client data generators.

LexManos opened the pull request:
Started off because of https://github.com/MinecraftForge/MinecraftForge/issues/10202

Main thing is the change to the launch handlers by adding 'forge_dev' and 'forge_userdev' which take in a '--launchEntry' argument to specify any main class.
This essentially removes the need for all other launch handlers, but they are kept for legacy/simplicity reasons.

Vanilla now exposes their client side data generators, notably they generate model and blockstate files. So I have marked all the data providers Forge adds for removal as modders should use vanilla. Less stuff for us to maintain. Namely everything in the `net.minecraftforge.client.model.generators` and `net.minecraftforge.client.model.generators.loaders` packages. Yes I know there are not vanilla equivalents for the Obj model stuff. But that should be as simple as a helper method instead of the full model generation system. So we can expand that if people want.

Vanilla now loads all block models even if they are not referenced by anything, and then it filters out those that are not referenced by another model/block. So I added a simple `ModelEvent.RegisterModelStateDefinitions` event to allow registration of synthetic block model states so that you can tell it to load custom models for blocks that are not registered in the main registry.

Jonathing commented:
I see that you've also deleted some GH Actions automations in this PR. Is it high on the priority list to re-implement them soon? Or nah?

Jonathing commented on fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/RenderElement.java:
```diff
@@ -84,16 +84,6 @@ private static void startupLogMessages(SimpleBufferBuilder bb, SimpleFont font,
 
         font.generateVerticesForTexts(10, context.scaledHeight() -  texts.size() * font.lineSpacing() + font.descent() - 10, bb, texts.toArray(SimpleFont.DisplayText[]::new));
     }
-    public static RenderElement monag() {
-        return new RenderElement(RenderElement.initializeTexture("monagstudios.png", 45000, 4, (bb, ctx, sz, frame) -> {
-            var size = 256;
-            var x0 = (ctx.width() - 2 * size) / 2;
-            var y0 = 64;
-            QuadHelper.loadQuad(bb, x0, x0+size, y0, y0+size/2f, 0f, 1f, 0f, 0.5f, 0xFFFFFFFF);
-            QuadHelper.loadQuad(bb, x0+size, x0+2*size, y0, y0+size/2f, 0f, 1f, 0.5f, 1f, 0xFFFFFFFF);
-        }));
-
-    }
```
Good riddance. Though, I think it'd be cool if some future implementation of the early display window could support extensions such as this.

As a reminder, in the future the best way to re-implement this would be a **separate window** from Minecraft that does not run on the render thread.

cc @basdxz

LexManos replied:
This just deletes a unused meme things that was added for some dumb reason. Not sure what you're talking about.

Jonathing replied:
For example, if a mod or modpack wants to add, replace, or remove images or text from the early display window. It's not something that would be necessary but is something that modders already do with hacks.

LexManos replied:
Its completely outside the scope of this pr and has nothing to do with the current code. So not the place.
And honestly I  would rather just delete the entire early loading screen because of how hacky it is.
But feel free to make a feature request in a different PR if you think it is something people want.

Jonathing replied:
I'll talk to Ven about it some time next year. Been thinking about it a decent amount.

Jonathing commented on build_forge.gradle:
The formatting changes in this file have me generally concerned for the formatting in this project as a whole. I get that it's not a big deal, but I'd rather find some day in the future where we can just churn out non-breaking formatting issues so that they don't clutter up unrelated PRs.

LexManos replied:
We've had this argument many times, lets not go over this again.
This has some basic formatting cleanups because the mix of using notepad and an IDE to edit the file.

Jonathing replied:
As long as it doesn't become in the problem in the future then I will look the other way.

Jonathing commented on fmlloader/build.gradle:
```diff
@@ -36,6 +37,7 @@ dependencies {
     implementation(libs.terminalconsoleappender)
     implementation(libs.jimfs)
 
+    // Needed because we have a custom log4j plugin, and they removed package scanning and require a data file to be generated
```
Unrelated, but could this be part of the reason why 1.18.2 suffered from the Log4J bump? See #10190

LexManos replied:
As far as I know, no because the bump is to a version where this feature is deprecated but functional. Not removed.

Jonathing commented on src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java:
```diff
@@ -16,7 +16,12 @@
  *
  * @see ModelProvider
  * @see ModelBuilder
+ *
+ * In 1.21.4 Mojang exposed their data generators for their models. So it should be feasible to just use theirs.
+ * If you find something lacking feel free to open a PR so that we can extend it.
+ * @deprecated Use Vanilla's providers {@link net.minecraft.client.data.models.ModelProvider}
  */
+@Deprecated(since = "1.21.4", forRemoval = true)
```
I can look through that myself. I've often times needed to write extensions to Forge's own model generators for my older projects. If it's not cumbersome to maintain, I'll send them your way with another PR sometime after this one is merged.

Jonathing commented on build_forge.gradle:
```diff
@@ -117,9 +117,9 @@ dependencies {
     bootstrap(libs.bundles.asm)   // Needed by securemodules
-    
+
     implementation(libs.jopt.simple) { version { strictly '5.0.4' } }
```
We HAVE to do something about this jopt simple library issue. I'm willing to investigate. What library is causing this? Would it be possible to use a gradle plugin to force this? If we have to tack it on to FG6, I can.

LexManos replied:
Actually, looking at it, we dont need this anymore. I think we've got all the transitives down, but that isnt related to this PR

Jonathing replied:
Ok. Let me know if anything changes so I can look into it.

Jonathing reviewed (commented):
I've left some general comments that are worth reading. Not a review of the PR.

Additional notes:
- Added Eclipse Apt
- Bumped Bootstrap from 2.1.6 to 2.1.7

LexManos commented:
> I see that you've also deleted some GH Actions automations in this PR. Is it high on the priority list to re-implement them soon? Or nah?

Right, I had that commit pending but havent pushed it because its not important.
But no, the actions never did anything and I deleted them cuz github emailed me basically saying "wtf are these for?"
We can re-add them at a later date when we github action the entire project. Or someone feels like making them actually do something useful.

LexManos committed:
Add hooks for using Vanilla's new Client data generator
Add new hook for custom block models that do not have real blocks backing them
Some workspace cleanup fixing log4j annotation processor

Jonathing commented:
> But no, the actions never did anything and I deleted them cuz github emailed me basically saying "wtf are these for?"
We can re-add them at a later date when we github action the entire project. Or someone feels like making them actually do something useful.

If they're not high on the priority list then it's not a problem. I see that Paint on the Discord wants to keep them for now, so that'll be fine. A personal goal for the project is to redo basically all of the automation only using what we actually need, like simple checks and labels.

Jonathing commented on build_forge.gradle:
```diff
@@ -227,6 +227,31 @@ patcher {
+                 '--mod', '.+',
+                 '--all',
+                 '--output', rootProject.file('src/test/generated/'),
+                 '--existing', sourceSets.main.resources.srcDirs[0],
+                 '--existing', sourceSets.test.resources.srcDirs[0],
```
Duplicate existing flag

LexManos replied:
Naw, main vs test

Jonathing commented on patches/minecraft/net/minecraft/client/data/models/ModelProvider.java.patch:
Could a decent number of methods in this patch be moved to an extension interface instead? `IModelProvider`? If not that is fine, but the patch surface on this is unnecessarily large.

LexManos replied:
Naw I dont think it's too big of an issue, the bulk of the patch is constructors. Which cant be moved to an interface. 
The only options are the 4 get methods which I don't think is worth it.
